docs/src/appendix/mast.md

```diff
@@ -0,0 +1,526 @@
+    /// string constants, de-duplicated and packed.
+    strings_count: VInt64,
+    /// The strings table.
+    ///
+    /// The actual string content is stored in `data`
+    strings: [StringRef; self.strings_count],
+    /// The total size of the trailing `data` section
+    ///
+    /// While this value can be computed by fully decoding the MAST, we
+    /// want to make it convenient to embed this format in another one,
+    /// so having this field allows skipping past the MAST data by only
+    /// decoding a couple of VInt64-encoded integers.
+    data_size: VInt64,
+    /// Represents that the remainder of the buffer are unstructured
+    /// bytes used to store the data for various fields that can have
+    /// extra data associated with them.
+    ///
+    /// This data must be smaller than 2^32 bytes.
+    data: [u8],
+}
+```
+
+The primary features to note here are:
+
+* The format can be detected in a byte stream (using the magic string)
+* The format is versioned, allowing us to evolve the format in a backwards-compatible fashion
+* We use variable-width integers in most cases to avoid a lot of wasted bytes
+* We use tables of fixed-width metadata to enable random access. This also has the effect of permitting a decoder to only decode what it needs, rather than the entire forest. For example, to decode a single tree in the forest. Lastly, the table approach enables us to de-duplicate entries in those tables where possible, e.g. the `strings` table.
+
+##### Node Info Table
+
+Entries in the `node_info` table are encoded as follows:
+
+```rust
+/// Represents an offset into the `data` section
+pub type DataOffset = u32;
+
+/// Represents an index of the `node_info` table
+pub type NodeIndex = u32;
+
+/// This type provides information that applies to all
+/// node types, but is not a complete definition of a node
+/// in most cases. Instead, that data must be decoded from
+/// the trailing data section.
+#[repr(C)]
+pub struct MastNodeInfo {
+    /// The node type that is encoded.
+    ty: MastNodeType,
+    /// The offset in the trailing data section of `EncodedMastForest`
+    /// where the data for this node starts. If this node has no data,
+    /// then the value will always be zero. Otherwise, the data should
+    /// be decoded starting at this offset, the specifics of which
+    /// depend on the node type
+    offset: DataOffset,
+    /// The merkle root for this node
+    ///
+    /// In the case of `MastNodeType::External`, this field reflects
+    /// the digest of the external node.
+    digest: Digest,
+}
+
+#[repr(u8)]
+pub enum MastNodeType {
+    /// For more efficient decoding, and because of the
+    /// frequency with which these node types appear, we
+    /// directly represent the child indices for `Join`,
+    /// `Split`, and `Loop` inline. This incurs 8 bytes
+    /// of padding for other node types, except `Block`
+    /// which takes advantage of this to also inline some
+    /// of its data
```

Comment:
Probably a premature optimization on my part, but I wonder if we should always encode `MastNodeType` as 8 bytes (rather than the current 9). Basically, the first 4 bits could be used to encode the variant, and then we could still encode 2 node indexes with 30 bits each. This does cap the maximum number of nodes at $2^30$, but I don't think that's a particularly big concern.

On the other hand, having `MastNodeType` aligned on 8-byte boundaries may be beneficial.